### ISSUE TITLE
Listbox: Invalid prop types for `onChange` when passing an array to `value` without `multiple`

### ISSUE THREAD

**Issue by JannikWibkerQC** [tailwindlabs/headlessui#2711]
**What package within Headless UI are you using?**

@headlessui/react

**What version of that package are you using?**

v1.7.17

**What browser are you using?**

Chrome

**Reproduction URL**

https://codesandbox.io/p/sandbox/nice-sunset-wzp69r

> Is this enough or should I create a GitHub repo as well?

**Describe your issue**

Passing an array (assuming `string[]` for simplicity) as `value` to `Listbox` (without `multiple`) has the following effect:
- type inference kicks in and infers the type of `onChange` to `(selected: string[]) => void`
- `value[0]` is used as the selected value (if the array is not empty; passing an empty array is the only possibility to get a listbox without any value being selected; passing undefined doesn't work as it results in a warning about the component switching from uncontrolled to controlled)
- when `onChange` fires it **doesn't honor** the inferred type, the argument is `string` instead of `string[]`

In the example this can be observed by checking the `console.log` output (`App.tsx#16`)

It would be great if either:
- the type would get inferred to `(selected: string) => void`
- `onChange` would be called with `[<value>]` instead of `<value>`

**Comment by RobinMalfait** [comment]
Hey thanks for your question!

What you are noticing, while a bit confusing, is the expected behaviour. One thing we asked ourselves is "Is the `multiple` prop needed, or can we infer it based on the `value` of the Combobox?". You could argue that we can infer it:

- If an array is passed -> multiple
- If a non-array is passed -> not multiple

There is one edge case that we have to keep in mind, and that is tuples.
In a lot of cases it's true that you will be dealing with simple strings, or with objects. But sometimes `tuples` are used as a data structure (for example as a simple result from `Object.entries({ ... })`). 

Let's say instead of using `{ label: "Website", url: "https://example.com" }` you want it to be stored as `["Website", "https://example.com"]` for whatever reason. In this case the data structure is a tuple, but to JavaScript it is still an array. Alternatively, if you have a tuple with different data types such as `["Count", 12]`, then inferring the value would be `string | number` which would be incorrect.

_We_ know that in this case the array is a tuple and is 1 unit of information, but we can't tell JavaScript the intent of this being a tuple instead of an array.

If we would infer the data type, (and infer that this is now in `multiple` mode) then you would have incorrect behaviour.

> value[0] is used as the selected value (if the array is not empty; passing an empty array is the only possibility to get a listbox without any value being selected; passing undefined doesn't work as it results in a warning about the component switching from uncontrolled to controlled)

In this case you can use `null` as a value to make it nullable: https://headlessui.com/react/combobox#allowing-empty-values

So in summary, if you _want_ an array to be used then you have to pass the `multiple` prop. If you just want an array for the default value, then I would use `null` instead such that we don't have to switch between data types.

Hope this helps!
